File: litho-rendercore/src/main/java/com/facebook/rendercore/renderunits/HostRenderUnit.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.rendercore.renderunits;

import static com.facebook.rendercore.RenderUnit.Extension.extension;
import static com.facebook.rendercore.RenderUnit.RenderType.VIEW;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.View.OnFocusChangeListener;
import android.view.View.OnLongClickListener;
import android.view.View.OnTouchListener;
import androidx.annotation.IntDef;
import androidx.annotation.Nullable;
import com.facebook.rendercore.HostView;
import com.facebook.rendercore.InterceptTouchHandler;
import com.facebook.rendercore.RenderUnit;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public class HostRenderUnit extends RenderUnit<HostView> {

  private static final int UNSET = -1;
  private static final int SET_FALSE = 0;
  private static final int SET_TRUE = 1;

  @IntDef({UNSET, SET_FALSE, SET_TRUE})
  @Retention(RetentionPolicy.SOURCE)
  @interface TriState {}

  private long mId;
  private @Nullable Drawable mBackground;
  private @Nullable Drawable mForeground;
  private int mLayerType = View.LAYER_TYPE_NONE;
  private @TriState int mClickable = UNSET;
  private boolean mEnabled = true;
  private boolean mFocusable;
  private boolean mFocusableInTouchMode;
  private OnFocusChangeListener mOnFocusChangeListener;
  private InterceptTouchHandler mInterceptTouchHandler;
  private OnLongClickListener mLongClickListener;
  private OnClickListener mClickListener;
  private OnTouchListener mOnTouchListener;

  public HostRenderUnit(long id) {
    super(VIEW);
    mId = id;
    addMountUnmountExtensions(
        extension(this, sBackgroundBindFunction),
        extension(this, sForegroundBindFunction),
        extension(this, sTouchHandlersBindFunction),
        extension(this, sLayerTypeBindFunction));
  }

  @Override
  public HostView createContent(Context c) {
    return new HostView(c);
  }

  @Override
  public long getId() {
    return mId;
  }

  @Nullable
  public Drawable getBackground() {
    return mBackground;
  }

  @Nullable
  public Drawable getForeground() {
    return mForeground;
  }

  public int getLayerType() {
    return mLayerType;
  }

  public @TriState int getClickable() {
    return mClickable;
  }

  public boolean isEnabled() {
    return mEnabled;
  }

  public OnTouchListener getOnTouchListener() {
    return mOnTouchListener;
  }

  public OnClickListener getOnClickListener() {
    return mClickListener;
  }

  public OnLongClickListener getOnLongClickListener() {
    return mLongClickListener;
  }

  public InterceptTouchHandler getOnInterceptTouchEvent() {
    return mInterceptTouchHandler;
  }

  public OnFocusChangeListener getOnFocusChangeListener() {
    return mOnFocusChangeListener;
  }

  public boolean isFocusable() {
    return mFocusable;
  }

  public boolean isFocusableInTouchMode() {
    return mFocusableInTouchMode;
  }

  public void setBackground(Drawable background) {
    mBackground = background;
  }

  public void setForeground(Drawable foreground) {
    mForeground = foreground;
  }

  public void setLayerType(int layerType) {
    mLayerType = layerType;
  }

  public void setClickable(boolean clickable) {
    mClickable = clickable ? SET_TRUE : SET_FALSE;
  }

  public void setEnabled(boolean enabled) {
    mEnabled = enabled;
  }

  public void setFocusable(boolean focusable) {
    mFocusable = focusable;
  }

  public void setFocusableInTouchMode(boolean focusableInTouchMode) {
    mFocusableInTouchMode = focusableInTouchMode;
  }

  public void setOnFocusChangeListener(OnFocusChangeListener onFocusChangeListener) {
    mOnFocusChangeListener = onFocusChangeListener;
  }

  public void setInterceptTouchHandler(InterceptTouchHandler interceptTouchHandler) {
    mInterceptTouchHandler = interceptTouchHandler;
  }

  public void setLongClickListener(OnLongClickListener longClickListener) {
    mLongClickListener = longClickListener;
  }

  public void setClickListener(OnClickListener clickListener) {
    mClickListener = clickListener;
  }

  public void setOnTouchListener(OnTouchListener onTouchListener) {
    mOnTouchListener = onTouchListener;
  }

  public static Binder<HostRenderUnit, HostView> sBackgroundBindFunction =
      new Binder<HostRenderUnit, HostView>() {
        @Override
        public boolean shouldUpdate(
            HostRenderUnit currentValue,
            HostRenderUnit newValue,
            Object currentLayoutData,
            Object nextLayoutData) {
          final Drawable currentBackground = currentValue.mBackground;
          final Drawable newBackground = newValue.mBackground;
          if (currentBackground == null) {
            return newBackground != null;
          } else {
            return newBackground != null && !currentBackground.equals(newBackground);
          }
        }

        @Override
        public void bind(
            Context context, HostView hostView, HostRenderUnit hostRenderUnit, Object layoutData) {
          hostView.setBackground(hostRenderUnit.getBackground());
        }

        @Override
        public void unbind(
            Context context, HostView hostView, HostRenderUnit hostRenderUnit, Object layoutData) {
          hostView.setBackground(null);
        }
      };

  public static Binder<HostRenderUnit, HostView> sForegroundBindFunction =
      new Binder<HostRenderUnit, HostView>() {
        @Override
        public boolean shouldUpdate(
            HostRenderUnit currentValue,
            HostRenderUnit newValue,
            Object currentLayoutData,
            Object nextLayoutData) {
          final Drawable currentForeground = currentValue.mForeground;
          final Drawable newForeground = newValue.mForeground;
          if (currentForeground == null) {
            return newForeground != null;
          } else {
            return newForeground != null && !currentForeground.equals(newForeground);
          }
        }

        @Override
        public void bind(
            Context context, HostView hostView, HostRenderUnit hostRenderUnit, Object layoutData) {
          hostView.setForegroundCompat(hostRenderUnit.getForeground());
        }

        @Override
        public void unbind(
            Context context, HostView hostView, HostRenderUnit hostRenderUnit, Object layoutData) {
          hostView.setForegroundCompat(null);
        }
      };

  public static Binder<HostRenderUnit, HostView> sLayerTypeBindFunction =
      new Binder<HostRenderUnit, HostView>() {

        @Override
        public boolean shouldUpdate(
            HostRenderUnit currentModel,
            HostRenderUnit newModel,
            @Nullable Object currentLayoutData,
            @Nullable Object nextLayoutData) {
          return currentModel.mLayerType != newModel.mLayerType;
        }

        @Override
        public void bind(
            Context context,
            HostView hostView,
            HostRenderUnit hostRenderUnit,
            @Nullable Object layoutData) {
          hostView.setLayerType(hostRenderUnit.getLayerType(), null);
        }

        @Override
        public void unbind(
            Context context,
            HostView hostView,
            HostRenderUnit hostRenderUnit,
            @Nullable Object layoutData) {
          hostView.setLayerType(View.LAYER_TYPE_NONE, null);
        }
      };

  public static Binder<HostRenderUnit, HostView> sTouchHandlersBindFunction =
      new Binder<HostRenderUnit, HostView>() {
        @Override
        public boolean shouldUpdate(
            HostRenderUnit currentValue,
            HostRenderUnit newValue,
            Object currentLayoutData,
            Object nextLayoutData) {
          // Updating touch and click listeners is not an expensive operation.
          return true;
        }

        @Override
        public void bind(
            Context context, HostView hostView, HostRenderUnit hostRenderUnit, Object layoutData) {
          hostView.setOnTouchListener(hostRenderUnit.getOnTouchListener());
          hostView.setInterceptTouchEventHandler(hostRenderUnit.getOnInterceptTouchEvent());
          final OnClickListener onClickListener = hostRenderUnit.getOnClickListener();
          if (onClickListener != null) {
            hostView.setOnClickListener(onClickListener);
          }

          final OnLongClickListener onLongClickListener = hostRenderUnit.getOnLongClickListener();
          if (onLongClickListener != null) {
            hostView.setOnLongClickListener(onLongClickListener);
          }

          OnFocusChangeListener onFocusChangeListener = hostRenderUnit.getOnFocusChangeListener();
          hostView.setOnFocusChangeListener(onFocusChangeListener);
          hostView.setFocusable(hostRenderUnit.isFocusable());
          hostView.setFocusableInTouchMode(hostRenderUnit.isFocusableInTouchMode());
          hostView.setEnabled(hostRenderUnit.isEnabled());

          final @TriState int clickable = hostRenderUnit.getClickable();
          if (clickable != UNSET) {
            hostView.setClickable(clickable == SET_TRUE);
          }
        }

        @Override
        public void unbind(
            Context context, HostView hostView, HostRenderUnit hostRenderUnit, Object layoutData) {
          hostView.setOnTouchListener(null);
          hostView.setInterceptTouchEventHandler(null);
          hostView.setOnClickListener(null);
          hostView.setClickable(false);
          hostView.setOnLongClickListener(null);
          hostView.setLongClickable(false);
          hostView.setOnFocusChangeListener(null);
          hostView.setFocusable(false);
          hostView.setFocusableInTouchMode(false);
        }
      };
}

File: litho-rendercore/src/main/java/com/facebook/rendercore/renderunits/HostRenderUnit.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.rendercore.renderunits

import android.content.Context
import android.graphics.drawable.Drawable
import android.view.View
import android.view.View.OnFocusChangeListener
import androidx.annotation.IntDef
import com.facebook.rendercore.ContentAllocator
import com.facebook.rendercore.HostView
import com.facebook.rendercore.RenderUnit
import com.facebook.rendercore.RenderUnit.DelegateBinder.Companion.createDelegateBinder

open class HostRenderUnit(override val id: Long) :
    RenderUnit<HostView>(RenderType.VIEW), ContentAllocator<HostView> {
  @IntDef(value = [UNSET, SET_FALSE, SET_TRUE])
  @Retention(AnnotationRetention.SOURCE)
  internal annotation class TriState

  var background: Drawable? = null
  var foreground: Drawable? = null
  var layerType: Int = View.LAYER_TYPE_NONE

  @get:TriState
  @TriState
  var clickable: Int = UNSET
    private set

  var isEnabled: Boolean = true
  var isFocusable: Boolean = false
  var isFocusableInTouchMode: Boolean = false
  var onFocusChangeListener: OnFocusChangeListener? = null
  var onClickListener: View.OnClickListener? = null

  override fun createContent(context: Context): HostView {
    return HostView(context)
  }

  override val contentAllocator: ContentAllocator<HostView>
    get() = this

  fun setClickable(clickable: Boolean) {
    this.clickable = if (clickable) SET_TRUE else SET_FALSE
  }

  init {
    addOptionalMountBinders(
        createDelegateBinder(this, backgroundBindFunction),
        createDelegateBinder(this, foregroundBindFunction),
        createDelegateBinder(this, touchHandlersBindFunction),
        createDelegateBinder(this, layerTypeBindFunction))
  }

  companion object {
    private const val UNSET = -1
    private const val SET_FALSE = 0
    private const val SET_TRUE = 1
    val backgroundBindFunction: Binder<HostRenderUnit, HostView, Any> =
        object : Binder<HostRenderUnit, HostView, Any> {
          override fun shouldUpdate(
              currentValue: HostRenderUnit,
              newValue: HostRenderUnit,
              currentLayoutData: Any?,
              nextLayoutData: Any?
          ): Boolean {
            val currentBackground = currentValue.background
            val newBackground = newValue.background
            return if (currentBackground == null) {
              newBackground != null
            } else {
              currentBackground != newBackground
            }
          }

          override fun bind(
              context: Context,
              hostView: HostView,
              hostRenderUnit: HostRenderUnit,
              layoutData: Any?
          ): Any? {
            hostView.background = hostRenderUnit.background
            return null
          }

          override fun unbind(
              context: Context,
              hostView: HostView,
              hostRenderUnit: HostRenderUnit,
              layoutData: Any?,
              bindData: Any?
          ) {
            hostView.background = null
          }
        }
    val foregroundBindFunction: Binder<HostRenderUnit, HostView, Any> =
        object : Binder<HostRenderUnit, HostView, Any> {
          override fun shouldUpdate(
              currentValue: HostRenderUnit,
              newValue: HostRenderUnit,
              currentLayoutData: Any?,
              nextLayoutData: Any?
          ): Boolean {
            val currentForeground = currentValue.foreground
            val newForeground = newValue.foreground
            return if (currentForeground == null) {
              newForeground != null
            } else {
              newForeground != null && currentForeground != newForeground
            }
          }

          override fun bind(
              context: Context,
              hostView: HostView,
              hostRenderUnit: HostRenderUnit,
              layoutData: Any?
          ): Any? {
            hostView.setForegroundCompat(hostRenderUnit.foreground)
            return null
          }

          override fun unbind(
              context: Context,
              hostView: HostView,
              hostRenderUnit: HostRenderUnit,
              layoutData: Any?,
              bindData: Any?
          ) {
            hostView.setForegroundCompat(null)
          }
        }
    val layerTypeBindFunction: Binder<HostRenderUnit, HostView, Any> =
        object : Binder<HostRenderUnit, HostView, Any> {
          override fun shouldUpdate(
              currentModel: HostRenderUnit,
              newModel: HostRenderUnit,
              currentLayoutData: Any?,
              nextLayoutData: Any?
          ): Boolean {
            return currentModel.layerType != newModel.layerType
          }

          override fun bind(
              context: Context,
              hostView: HostView,
              hostRenderUnit: HostRenderUnit,
              layoutData: Any?
          ): Any? {
            hostView.setLayerType(hostRenderUnit.layerType, null)
            return null
          }

          override fun unbind(
              context: Context,
              hostView: HostView,
              hostRenderUnit: HostRenderUnit,
              layoutData: Any?,
              bindData: Any?
          ) {
            hostView.setLayerType(View.LAYER_TYPE_NONE, null)
          }
        }
    val touchHandlersBindFunction: Binder<HostRenderUnit, HostView, Any> =
        object : Binder<HostRenderUnit, HostView, Any> {
          override fun shouldUpdate(
              currentValue: HostRenderUnit,
              newValue: HostRenderUnit,
              currentLayoutData: Any?,
              nextLayoutData: Any?
          ): Boolean {
            // Updating touch and click listeners is not an expensive operation.
            return true
          }

          override fun bind(
              context: Context,
              hostView: HostView,
              hostRenderUnit: HostRenderUnit,
              layoutData: Any?
          ): Any? {
            val onClickListener = hostRenderUnit.onClickListener
            if (onClickListener != null) {
              hostView.setOnClickListener(onClickListener)
            }
            val onFocusChangeListener = hostRenderUnit.onFocusChangeListener
            hostView.onFocusChangeListener = onFocusChangeListener
            hostView.isFocusable = hostRenderUnit.isFocusable
            hostView.isFocusableInTouchMode = hostRenderUnit.isFocusableInTouchMode
            hostView.isEnabled = hostRenderUnit.isEnabled
            @TriState val clickable = hostRenderUnit.clickable
            if (clickable != UNSET) {
              hostView.isClickable = clickable == SET_TRUE
            }
            return null
          }

          override fun unbind(
              context: Context,
              hostView: HostView,
              hostRenderUnit: HostRenderUnit,
              layoutData: Any?,
              bindData: Any?
          ) {
            hostView.setOnClickListener(null)
            hostView.isClickable = false
            hostView.onFocusChangeListener = null
            hostView.isFocusable = false
            hostView.isFocusableInTouchMode = false
          }
        }
  }
}
